core/include/gnuradio-4.0/TensorMath.hpp

```diff
@@ -0,0 +1,785 @@
+        TensorType result(tensor);
+        return multiply_scalar_inplace(result, scalar);
+    }
+
+    [[maybe_unused]] static constexpr TensorType& divide_scalar_inplace(TensorType& self, const T& scalar) {
+        if (scalar == T{0}) {
+            throw std::runtime_error("Division by zero");
+        }
+#if defined(__GLIBCXX__)
+        std::transform(std::execution::unseq, self.begin(), self.end(), self.begin(), [scalar](const T& x) { return x / scalar; });
+#else
+        std::ranges::transform(self.begin(), self.end(), self.begin(), [scalar](const T& x) { return x / scalar; });
+#endif
+        return self;
+    }
+
+    [[nodiscard]] static constexpr TensorType divide_scalar(const TensorType& tensor, const T& scalar) {
+        TensorType result(tensor);
+        return divide_scalar_inplace(result, scalar);
+    }
+
+    // element-wise multiplication (Hadamard product)
+    [[maybe_unused]] static constexpr TensorType& multiply_elementwise_inplace(TensorType& self, const TensorType& other) {
+        auto same_shape = [](std::span<const std::size_t> a1, std::span<const std::size_t> b1) { return a1.size() == b1.size() && std::equal(a1.begin(), a1.end(), b1.begin()); };
+        if (!same_shape(self.extents(), other.extents())) {
+            throw std::runtime_error("Tensor dimensions must match for element-wise operations");
+        }
+
+#if defined(__GLIBCXX__)
+        std::ranges::transform(std::execution::unseq, self, other.begin(), self.begin(), std::multiplies<>{});
+#else
+        std::ranges::transform(self, other.begin(), self.begin(), std::multiplies<>{});
+#endif
+        return self;
+    }
+
+    [[nodiscard]] static constexpr TensorType multiply_elementwise(const TensorType& a, const TensorType& b) {
+        TensorType result(a);
+        return multiply_elementwise_inplace(result, b);
+    }
+
+    [[maybe_unused]] static constexpr TensorType& divide_elementwise_inplace(TensorType& self, const TensorType& other) {
+        auto same_shape = [](std::span<const std::size_t> a1, std::span<const std::size_t> b1) { return a1.size() == b1.size() && std::equal(a1.begin(), a1.end(), b1.begin()); };
+        if (!same_shape(self.extents(), other.extents())) {
+            throw std::runtime_error("Tensor dimensions must match for element-wise operations");
+        }
+#if defined(__GLIBCXX__)
+        std::ranges::transform(std::execution::unseq, self, other.begin(), self.begin(), std::divides<>{});
+#else
+        std::ranges::transform(self, other.begin(), self.begin(), std::divides<>{});
+#endif
+        return self;
+    }
+
+    [[nodiscard]] static constexpr TensorType divide_elementwise(const TensorType& a, const TensorType& b) {
+        TensorType result(a);
+        return divide_elementwise_inplace(result, b);
+    }
+
+    [[nodiscard]] static constexpr T sum(const TensorType& self) { return std::accumulate(self.begin(), self.end(), T{0}); }
```

Comment:
Any reason for using accumulate here, and reduce in product? Also, would `unseq` for reduce make a difference here?